modules/meta/arrow-meta-prototype/testing-plugin/src/main/kotlin/arrow/meta/plugin/testing/Compilation.kt

```diff
@@ -0,0 +1,65 @@
+package arrow.meta.plugin.testing
+
+import com.tschuchort.compiletesting.KotlinCompilation
+import com.tschuchort.compiletesting.SourceFile
+import org.assertj.core.api.Assertions.assertThat
+import java.io.File
+import java.nio.file.Paths
+import io.github.classgraph.ClassGraph
+
+enum class CompilationResult {
+  OK,
+  INTERNAL_ERROR,
+  COMPILATION_ERROR,
+  SCRIPT_EXECUTION_ERROR
+}
+
+data class CompilationData(
+  val sourceFileName: String,
+  val sourceContent: String,
+  val generatedFileContent: String,
+  val generatedClasses: ArrayList<String>,
+  val compilationResult: CompilationResult
+)
+
+fun testCompilation(compilationData: CompilationData) {
+  val kotlinSource = SourceFile.kotlin(compilationData.sourceFileName, compilationData.sourceContent)
+
+  val result = KotlinCompilation().apply {
+    sources = listOf(kotlinSource)
+    classpaths = listOf(
+      // TODO
+      // classpathOf("arrow-annotations:x.x.x")
+      File("../../arrow-annotations/build/libs/arrow-annotations-0.10.1-SNAPSHOT.jar")
+    )
+    pluginClasspaths = listOf(
+      classpathOf("compiler-plugin")
+//      File("../compiler-plugin/build/libs/compiler-plugin.jar")
+    )
+  }.compile()
+
+  assertThat(result.exitCode).isEqualTo(exitCodeFrom(compilationData.compilationResult))
+
+  if (!result.exitCode.equals(KotlinCompilation.ExitCode.OK)) return
```

Comment:
Early returns like this are great for performance but I think in this case is better to take the bottom of the function remaining and splitting in a new function that is conditionally applied based on the value of the exit code.

Reply:
I love that idea! Thanks!